File: ios/Samples/AndorraTelecomPublic/AndorraTelecomPublic/Views/CallMeBackView.swift
/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import UIKit
import Alamofire

@objcMembers
class CallMeBackView: UIView {

    weak var delegate: CallMeBackDelegate?

    @IBOutlet var view: UIView!

    @IBOutlet weak var title: UILabel!
    @IBOutlet weak var phoneLabel: UILabel!
    @IBOutlet weak var phoneTextField: UITextFieldExtended!
    @IBOutlet weak var legalLabel: UILabel!
    @IBOutlet weak var legalSwitch: UISwitch!
    @IBOutlet weak var callMeNowButton: UIButton!
    @IBOutlet weak var orLabel: UILabel!
    @IBOutlet weak var iCallButton: UIButton!

    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)

        Bundle.main.loadNibNamed("CallMeBackView", owner: self, options: nil)
        self.addSubview(view)
        view.frame = self.bounds

        setTextOutlets()
        addTouchUpToLegalLabel()
    }

    func addTouchUpToLegalLabel() {
        legalLabel.isUserInteractionEnabled = true
        let tapGesture = UITapGestureRecognizer(target: self, action: #selector(legalTouchUpPopOver(tapGestureRecognizer:)))
        legalLabel.addGestureRecognizer(tapGesture)
    }

    func legalTouchUpPopOver(tapGestureRecognizer: UITapGestureRecognizer) {
        delegate?.showLegalCoditions(callMeBackView: self)
    }

    func legalConditionsChange(isAccepted: Bool) {
        legalSwitch.setOn(isAccepted, animated: true)
    }

    @IBAction func callMeNow(_ sender: UIButton) {
        let phoneCount = phoneTextField.text?.count
        if phoneCount != 6 {
            delegate?.showAlertLegalNotAccepted(callMeBackView: self, title: "title-legal".localized()!, message: "phone-not-valid".localized()!)
        }
        if legalSwitch.isOn {
            delegate?.showAlertLegalNotAccepted(callMeBackView: self, title: "title-call".localized()!, message: "call".localized()!)
        }
        else {
            delegate?.showAlertLegalNotAccepted(callMeBackView: self, title: "title-legal".localized()!, message: "legal-no-accept".localized()!)
        }
    }

    @IBAction func iCall(_ sender: UIButton) {
        if let url = URL(string: "tel://100900900") {
            if #available(iOS 10, *) {
                UIApplication.shared.open(url, options: [:], completionHandler: nil)
            }
            else {
                UIApplication.shared.openURL(url as URL)
            }
        }
    }

    func setTextOutlets() {
        title.text = "help?".localized()
        phoneLabel.text = "phone".localized()

        let mutableString = NSMutableAttributedString(string: "accept-legal".localized()!, attributes: nil)
        let range = rangeLastTwoWords(words: "accept-legal".localized()!)
        let attributtes =
            [.foregroundColor: UIColor(red:0.83, green:0.02, blue:0.45, alpha:1.0),
            .underlineStyle: 1] as [NSAttributedStringKey : Any]
        mutableString.addAttributes(attributtes, range: range)

        legalLabel.attributedText = mutableString

        callMeNowButton.setTitle("call-me-now".localized(), for: .normal)
        orLabel.text = "or".localized()
        iCallButton.setTitle("i-call".localized(), for: .normal)
    }

    func rangeLastTwoWords(words: String) -> NSRange {
        let countWords = words.characters.count
        let arrayWords = words.components(separatedBy: " ")
        let lastPosition = arrayWords.count - 1

        let countLastTwoWords = arrayWords[lastPosition].characters.count + arrayWords[lastPosition - 1].characters.count + 1

        return NSRange(location: countWords - countLastTwoWords, length: countLastTwoWords)
    }
}

File: xamarin/Samples/AndorraTelecom-iOS/CustomViews/CallMeBackView.cs
using System;
using AndorraTelecomiOS.Util;
using Foundation;
using ObjCRuntime;
using UIKit;

namespace AndorraTelecomiOS
{
    public partial class CallMeBackView : UIView
    {
        public CallMeBackView(IntPtr handle) : base(handle) { }

        public static ICallMeBackDelegate Delegate;

        public static CallMeBackView Create()
        {
            var CallMeBackViewArr = NSBundle.MainBundle.LoadNib("CallMeBackView", null, null);
            var View = Runtime.GetNSObject<CallMeBackView>(CallMeBackViewArr.ValueAt(0));

            /* Set text outlets */

            var LanguageBundle = RetrieveLanguageBundle(LanguageHelper.Language);

            View.HelpQuestion.Text = LanguageBundle.LocalizedString("Help?", null);

            View.PhoneLabel.Text = LanguageBundle.LocalizedString("Phone", null);

            var PhoneImage = new UIImage("img/phone-input.png");
            View.PhoneInput.LeftView = new UIImageView(PhoneImage);
            View.PhoneInput.LeftViewMode = UITextFieldViewMode.Always;

            var Attributes = new UIStringAttributes();
            Attributes.UnderlineStyle = NSUnderlineStyle.Single;
            Attributes.ForegroundColor = Colors.DarkPurple;

            var AcceptLegalLocalized = LanguageBundle.LocalizedString("Accept-legal", null);
            var MutableString = new NSMutableAttributedString(AcceptLegalLocalized);
            var Range = RangeLastTwoWords(AcceptLegalLocalized);
            MutableString.AddAttributes(Attributes, Range);
            View.LegalLabel.AttributedText = MutableString;

            View.CallMeNowButton.SetTitle(LanguageBundle.LocalizedString("Call-me-now", null), UIControlState.Normal);
            View.OrLabel.Text = LanguageBundle.LocalizedString("Or", null);
            View.ICallButton.SetTitle(LanguageBundle.LocalizedString("I-call", null), UIControlState.Normal);

            /* Add gesture recognizer to legal conditions link */

            Action LegalTouchPopOver = () =>
            {
                ShowLegalConditions(View);
            };

            var TapGesture = new UITapGestureRecognizer(LegalTouchPopOver);
            View.LegalLabel.UserInteractionEnabled = true;
            View.LegalLabel.AddGestureRecognizer(TapGesture);

            return View;
        }

        /* Actions */

        partial void CallMeNowButton_TouchUpInside(UIButton sender)
        {
            var LanguageBundle = RetrieveLanguageBundle(LanguageHelper.Language);

            var PhoneCount = PhoneInput.Text.Length;
            if(PhoneCount != 6)
            {
                ShowAlertLegalNotAccepted(this, LanguageBundle.LocalizedString("Title-invalid-number", null), LanguageBundle.LocalizedString("Phone-not-valid", null));
            }
            else if(LegalSwitch.On)
            {
                ShowAlertLegalNotAccepted(this, LanguageBundle.LocalizedString("Title-call", null), LanguageBundle.LocalizedString("Call", null));
            }
            else
            {
                ShowAlertLegalNotAccepted(this, LanguageBundle.LocalizedString("Title-legal", null), LanguageBundle.LocalizedString("Legal-not-accepted", null));
            }
        }

        partial void ICallButton_TouchUpInside(UIButton sender)
        {
            var Url = new NSUrl("tel://100900900");
            if (UIApplication.SharedApplication.CanOpenUrl(Url))
            {
                UIApplication.SharedApplication.OpenUrl(Url);
            }
            else
            {
                Console.WriteLine("Can't open Url: " + Url);
            }
        }

        /* Private methods */

        static NSBundle RetrieveLanguageBundle(string Language)
        {
            var Path = NSBundle.MainBundle.PathForResource(Language, "lproj");
            return NSBundle.FromPath(Path);
        }

        static NSRange RangeLastTwoWords(string Message)
        {
            var CountCharacters = Message.Length;
            var ArrayWords = Message.Split(' ');
            var LastPosition = ArrayWords.Length - 1;

            var CountLastTwoWords = ArrayWords[LastPosition].Length + ArrayWords[LastPosition - 1].Length + 1;

            return new NSRange(CountCharacters - CountLastTwoWords, CountLastTwoWords);
        }

        static void ShowLegalConditions(CallMeBackView View)
        {
            Delegate.ShowLegalConditions(View);
        }

        static void ShowAlertLegalNotAccepted(CallMeBackView View, string Title, string Message)
        {
            Delegate.ShowAlertLegalNotAccepted(View, Title, Message);
        }

        internal void LegalConditionsChange(CallMeBackView View, bool IsAccepted)
        {
            View.LegalSwitch.On = IsAccepted;
        }
    }
}